### PR TITLE
feat: Add invoked measurement.

### PR DESCRIPTION
**Pull request by kinyoklion**
The migration_op event now has a measurement for which "origins" were invoked. Basically was the 'new' method, the 'old' method, or both executed.

This is because from other data you cannot safely infer what was executed. We need to know total executions in order to properly analyze results.

### CODE REVIEW

File: contract-tests/index.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ app.get('/', (req, res) => {
       'big-segments',
       'user-type',
       'migrations',
+      'event-sampling',
```
Comment (kinyoklion) [inline contract-tests/index.js:32]: To opt-in to the event-sampling tests in the contract test.

File: contract-tests/sdkClientEntity.js [hunk 2]
```diff
@@ -64,7 +64,7 @@ export function makeSdkConfig(options, tag) {
 }
```
Comment (kinyoklion) [inline contract-tests/sdkClientEntity.js:64]: Prettier attacked this file.

File: packages/shared/sdk-server/__tests__/MigrationOpEvent.test.ts [hunk 3]
```diff
@@ -91,9 +91,9 @@ describe('given an LDClient with test data', () => {
           // Migration event.
           const migrationEvent = (await events.take()) as internal.InputMigrationEvent;
           // Only check the measurements component of the event.
-          expect(migrationEvent.measurements[0].key).toEqual('consistent');
+          expect(migrationEvent.measurements[1].key).toEqual('consistent');
```
Comment (kinyoklion) [inline packages/shared/sdk-server/__tests__/MigrationOpEvent.test.ts:94]: Adding the new measurement consistently shifted things.

File: packages/shared/sdk-server/src/MigrationOpTracker.ts [hunk 4]
```diff
@@ -56,31 +63,118 @@ export default class MigrationOpTracker implements LDMigrationTracker {
...
+        default: this.defaultStage,
+        reason: this.reason,
+        variation: this.variation,
+      },
+      measurements,
+      samplingRatio: this.samplingRatio ?? 1,
+    };
+  }
+
+  private logTag(): string {
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/MigrationOpTracker.ts:116]: I added various log messages for things that an be wrong with an event. These basically only apply if you are using the blocks but not the kit.

**Comment by kinyoklion** [comment]
Tests are fixed in the PR which adds the new data kinds to edge SDKs.

File: packages/shared/sdk-server/src/Migration.ts [hunk 5]
```diff
@@ -302,6 +302,7 @@ export default class Migration<
     origin: LDMigrationOrigin,
     method: (payload?: TInput) => Promise<LDMethodResult<TOutput>>,
   ): Promise<LDMigrationResult<TOutput>> {
+    context.tracker.invoked(origin);
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/Migration.ts:305]: This is the call that actually causes the tracker method to be invoked in all cases.

File: packages/shared/sdk-server/src/MigrationOpTracker.ts [hunk 6]
```diff
@@ -56,31 +63,118 @@ export default class MigrationOpTracker implements LDMigrationTracker {
...
+    }
+
+    if (Object.keys(this.contextKeys).length === 0) {
+      this.logger?.error(
+        'The migration was not done against a valid context and cannot generate an event.',
+      );
+      return undefined;
+    }
+
+    if (!this.wasInvoked.old && !this.wasInvoked.new) {
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/MigrationOpTracker.ts:83]: If neither the old or the new implementation was invoked, then that isn't valid, and we will not create the event.

File: packages/shared/sdk-server/src/MigrationOpTracker.ts [hunk 7]
```diff
@@ -20,6 +21,11 @@ export default class MigrationOpTracker implements LDMigrationTracker {
     new: false,
   };
 
+  private wasInvoked = {
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/MigrationOpTracker.ts:24]: Used to track which implementations were executed.

File: packages/shared/sdk-server/src/MigrationOpTracker.ts [hunk 8]
```diff
@@ -38,6 +44,7 @@ export default class MigrationOpTracker implements LDMigrationTracker {
     private readonly checkRatio?: number,
     private readonly variation?: number,
     private readonly samplingRatio?: number,
+    private readonly logger?: LDLogger,
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/MigrationOpTracker.ts:47]: Added the logger so we can provide some useful information for invalid cases. Tracking what method was invoked allows us to perform some consistency checks on the data and log when there is an issue.

File: packages/shared/sdk-server/src/MigrationOpTracker.ts [hunk 9]
```diff
@@ -56,31 +63,118 @@ export default class MigrationOpTracker implements LDMigrationTracker {
...
-          key: this.flagKey,
-          value: this.stage,
-          default: this.defaultStage,
-          reason: this.reason,
-          variation: this.variation,
-        },
-        measurements,
-        samplingRatio: this.samplingRatio ?? 1,
-      };
+    if (!this.operation) {
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/MigrationOpTracker.ts:71]: Inverted the logic here to have early returns.

**Review by yusinto — APPROVED** [review]

File: packages/shared/sdk-server/src/MigrationOpEventConversion.ts [hunk 10]
```diff
@@ -28,6 +29,12 @@ function isErrorMeasurement(value: LDMigrationMeasurement): value is LDMigration
   return (value as any).kind === undefined && value.key === 'error';
 }
 
+function isInvokedMeasurement(
+  value: LDMigrationMeasurement,
+): value is LDMigrationInvokedMeasurement {
+  return (value as any).kind === undefined && value.key === 'invoked';
```
[the remark this answers is not in the document]
Comment (kinyoklion) [reply packages/shared/sdk-server/src/MigrationOpEventConversion.ts:35]: I can remove that. Previously we were going to support `custom` measurements. Which had a kind, which created an unpleasant difference in shape. 

They have been tabled for now though, so this isn't needed.

**Pull request merged**